conmon-rs/server/src/attach.rs

```diff
@@ -46,43 +102,125 @@ impl Attach {
         permissions.set_mode(0o700);
 
         listen(fd, 10).context("listen on socket fd")?;
-        thread::spawn(move || Self::start_listening(fd));
 
-        Ok(Self {})
+        let clients = Arc::new(RwLock::new(vec![]));
+        let clients_clone = clients.clone();
+        task::spawn(async move { Self::start_listening(fd, clients_clone).await });
+
+        Ok(Self {
+            clients,
+            path: socket_path.into(),
+        })
     }
 
-    fn start_listening(fd: RawFd) -> Result<()> {
+    async fn start_listening(fd: RawFd, clients: Clients) -> Result<()> {
         debug!("Start listening on attach socket");
-        let listener = unsafe { UnixListener::from_raw_fd(fd) };
-        for stream in listener.incoming() {
-            thread::spawn(|| Self::handle_client(stream?));
+        let listener = UnixListener::from_std(unsafe { net::UnixListener::from_raw_fd(fd) })?;
+        loop {
+            match listener.accept().await {
+                Ok((stream, _)) => {
+                    debug!("Got new attach stream connection");
+                    clients.write().await.push(stream);
+                }
+                Err(e) => error!("Unable to accept attach stream: {}", e),
+            }
         }
-        Ok(())
     }
 
-    fn handle_client(stream: UnixStream) -> Result<()> {
-        debug!("Got new attach stream connection");
+    /// Try to read from all streams standard input and return the first result.
+    pub async fn try_read(&self) -> Result<Option<Vec<u8>>> {
+        let mut cleanup_idxs = vec![];
+        let mut clients = self.clients.write().await;
 
-        let mut reader = BufReader::new(&stream);
-        let mut writer = BufWriter::new(&stream);
+        for (idx, stream) in clients.iter().enumerate() {
```

Comment:
this will mean if we've registered multiple clients, the first will get priority for stdin, right? FYI @mheon for podman integration

Reply:
Yes that's correct.